src/libraries/System.Security.Cryptography.Encoding/src/System/Security/Cryptography/PemEncoding.cs

```diff
@@ -0,0 +1,435 @@
+        /// </exception>
+        /// <returns>
+        /// A <see cref="PemFields"/> structure that contains the location, label, and
+        /// data location of the encoded data.
+        /// </returns>
+        public static PemFields Find(ReadOnlySpan<char> pemData)
+        {
+            if (!TryFind(pemData, out PemFields fields))
+            {
+                throw new ArgumentException(SR.Argument_PemEncoding_NoPemFound, nameof(pemData));
+            }
+            return fields;
+        }
+
+        /// <summary>
+        /// Finds the first PEM-encoded data.
+        /// </summary>
+        /// <param name="pemData">
+        /// A span containing the PEM encoded data.
+        /// </param>
+        /// <param name="fields">
+        /// When this method returns <c>true</c>, the found <see cref="PemFields"/> structure
+        /// that contains the location, label, and data location of the encoded data.
+        /// </param>
+        /// <returns>
+        /// <c>true</c> if PEM encoded data was found; otherwise <c>false</c>.
+        /// </returns>
+        public static bool TryFind(ReadOnlySpan<char> pemData, out PemFields fields)
+        {
+            // Check for the minimum possible encoded length of a PEM structure
+            // and exit early if there is no way the input could contain a well-formed
+            // PEM.
+            if (pemData.Length < Preeb.Length + Ending.Length * 2 + Posteb.Length)
+            {
+                fields = default;
+                return false;
+            }
+
+            Span<char> postebShortBuffer = stackalloc char[256];
+            int preebLinePosition = 0;
+            while (TryReadNextLine(pemData, ref preebLinePosition, out Range lineRange))
+            {
+                ReadOnlySpan<char> line = pemData[lineRange];
+                int preebIndex = line.IndexOf(Preeb);
+
+                if (preebIndex == -1 ||
+                    (preebIndex > 0 && !line[..preebIndex].IsWhiteSpace())) // can only be preceeded by whitespace
+                {
+                    continue;
+                }
+
+                int preebEndingIndex = line[(preebIndex + Preeb.Length)..].IndexOf(Ending);
+
+                if (preebEndingIndex == -1)
+                {
+                    continue;
+                }
+
+                (int preebOffset, _) = lineRange.GetOffsetAndLength(pemData.Length);
+                int preebStartIndex = preebOffset + preebIndex;
+                int startLabelIndex = preebStartIndex + Preeb.Length;
+                int endLabelIndex = startLabelIndex + preebEndingIndex;
+                Range labelRange = startLabelIndex..endLabelIndex;
+                ReadOnlySpan<char> label = pemData[labelRange];
+
+                if (!IsValidLabel(label))
+                {
+                    continue;
+                }
+
+                int postebLength = Posteb.Length + label.Length + Ending.Length;
```

Comment:
Could this overflow?

Reply:
I don't think so. `label`'s length is going to be based on a found label in the PEM, and we've already checked for the `"-----BEGIN "` and `"-----"`, so we know that the ROS (whose max length is an Int32) contains something of this length.